File: internal/enforcement/judge.go
package enforcement

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// EvaluateInput bundles everything the judge LLM needs to reason about
// a request. Bundled rather than positional so adding context fields
// later doesn't churn every caller.
type EvaluateInput struct {
	CageType     string
	VulnClass    string
	AssessmentID string
	Method       string
	URL          string
	Headers      map[string]string
	Body         []byte
	Objective    string
	AgentReason  string
}

// EvaluateOutput bundles the judge's decision, its reason, and any
// usage reported by the provider. Returned as a struct so adding
// fields later (e.g. provider, latency) doesn't churn every caller.
type EvaluateOutput struct {
	Decision PayloadDecision
	Reason   string
	Usage    JudgeUsage
}

// JudgeUsage carries token counts from the judge LLM call so the
// payload-proxy can count judge spend against the cage's token budget
// alongside the agent's direct LLM calls.
type JudgeUsage struct {
	PromptTokens     int64
	CompletionTokens int64
	TotalTokens      int64
}

type JudgeClient struct {
	endpoint            string
	confidenceThreshold float64
	apiKey              string
	httpClient          *http.Client
}

func NewJudgeClient(endpoint string, confidenceThreshold float64, apiKey string, timeout time.Duration) *JudgeClient {
	return &JudgeClient{
		endpoint:            endpoint,
		confidenceThreshold: confidenceThreshold,
		apiKey:              apiKey,
		httpClient:          &http.Client{Timeout: timeout},
	}
}

// SetTransport overrides the HTTP transport used for judge requests.
// The payload proxy uses this to set the fwmark transport so judge
// connections bypass the iptables redirect.
func (c *JudgeClient) SetTransport(t http.RoundTripper) {
	c.httpClient.Transport = t
}

// The judge prompt and tool schema live here (not in the webhook).
// The webhook is a dumb forwarder of OpenAI chat completion requests;
// the response shape is structurally enforced by the tool-use schema
// below, so an operator can rewrite our system prompt in their own
// custom webhook without breaking the wire contract — as long as they
// keep the tool definition intact.

const judgeSystemPrompt = `You are a payload-safety judge for an authorized penetration testing system. You evaluate whether a specific HTTP request the cage agent wants to send is safe to forward to the target. The cage is sandboxed and the target is explicitly authorized.

SAFE (allow forwarding):
- Read-only probes (GET requests, status checks, header inspections)
- Authentication attempts with low-volume default credentials against ONE endpoint
- Canary or boundary-test injections (XSS canary, basic SQL syntax probe) that don't mutate state
- Most state-changing requests within authorized pentest scope

UNSAFE (block forwarding):
- Mass destruction (DROP TABLE, DELETE without WHERE, account purge endpoints)
- Credential exfiltration intended for reuse on live systems
- Lateral movement outside the authorized target scope
- POSTs to attacker-controlled endpoints that leak target data
- Anything obviously catastrophic and irreversible

Lean toward SAFE unless harm is obvious — over-blocking defeats the pentest's purpose. The agent has the operator's permission.

Call the submit_judgment function with your decision. Confidence reflects how certain you are: 1.0 = obvious case, 0.5 = genuinely uncertain.`

// judgeToolSchema is the OpenAI function-calling schema agentcage
// requires the judge LLM to emit. The provider enforces this at the
// inference layer (with tool_choice: "required"), so even if an
// operator rewrites the system prompt in their custom webhook, the
// shape of the response is locked.
var judgeToolSchema = map[string]any{
	"type": "function",
	"function": map[string]any{
		"name":        "submit_judgment",
		"description": "Return your safety judgment for the request.",
		"parameters": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"safe":       map[string]any{"type": "boolean", "description": "Whether the request is safe to forward."},
				"confidence": map[string]any{"type": "number", "minimum": 0, "maximum": 1, "description": "How certain you are, 0.0 to 1.0."},
				"reason":     map[string]any{"type": "string", "description": "One sentence explaining the decision."},
			},
			"required": []string{"safe", "confidence", "reason"},
		},
	},
}

// Evaluate sends an OpenAI-format chat completion to the judge
// endpoint and returns a decision. The endpoint is a passthrough to an
// OpenAI-compatible provider (our shipped webhook is ~20 lines; an
// operator can fork it and transform requests for custom routing /
// extra context / provider switching). Uses its own timeout rather
// than the caller's context so the agent's request deadline cannot cut
// the judge call short.
func (c *JudgeClient) Evaluate(in EvaluateInput) (EvaluateOutput, error) {
	ctx, cancel := context.WithTimeout(context.Background(), c.httpClient.Timeout)
	defer cancel()

	userContent := formatJudgeUserMessage(in)

	reqBody, err := json.Marshal(map[string]any{
		"messages": []map[string]string{
			{"role": "system", "content": judgeSystemPrompt},
			{"role": "user", "content": userContent},
		},
		"tools":       []any{judgeToolSchema},
		"tool_choice": "required",
		"temperature": 0,
	})
	if err != nil {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge request marshal failed"}, fmt.Errorf("marshaling judge request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.endpoint, bytes.NewReader(reqBody))
	if err != nil {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge request creation failed"}, fmt.Errorf("creating judge request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("x-api-key", c.apiKey)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge unreachable"}, fmt.Errorf("calling judge endpoint: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode >= 400 {
		return EvaluateOutput{Decision: PayloadHold, Reason: fmt.Sprintf("judge returned status %d", resp.StatusCode)}, fmt.Errorf("judge endpoint returned %d", resp.StatusCode)
	}

	var chatResp struct {
		Choices []struct {
			Message struct {
				ToolCalls []struct {
					Function struct {
						Name      string `json:"name"`
						Arguments string `json:"arguments"`
					} `json:"function"`
				} `json:"tool_calls"`
			} `json:"message"`
		} `json:"choices"`
		Usage struct {
			PromptTokens     int64 `json:"prompt_tokens"`
			CompletionTokens int64 `json:"completion_tokens"`
			TotalTokens      int64 `json:"total_tokens"`
		} `json:"usage"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge response malformed"}, fmt.Errorf("decoding judge response: %w", err)
	}

	usage := JudgeUsage{
		PromptTokens:     chatResp.Usage.PromptTokens,
		CompletionTokens: chatResp.Usage.CompletionTokens,
		TotalTokens:      chatResp.Usage.TotalTokens,
	}

	if len(chatResp.Choices) == 0 || len(chatResp.Choices[0].Message.ToolCalls) == 0 {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge returned no tool call (provider may not support tool-use or operator stripped the tool definition)", Usage: usage}, fmt.Errorf("no tool_calls in judge response")
	}

	call := chatResp.Choices[0].Message.ToolCalls[0]
	if call.Function.Name != "submit_judgment" {
		return EvaluateOutput{Decision: PayloadHold, Reason: fmt.Sprintf("judge called unexpected function %q", call.Function.Name), Usage: usage}, fmt.Errorf("unexpected function %q", call.Function.Name)
	}

	var args struct {
		Safe       bool    `json:"safe"`
		Confidence float64 `json:"confidence"`
		Reason     string  `json:"reason"`
	}
	if err := json.Unmarshal([]byte(call.Function.Arguments), &args); err != nil {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge tool arguments not valid JSON", Usage: usage}, fmt.Errorf("decoding judge tool arguments: %w", err)
	}
	if args.Confidence < 0 || args.Confidence > 1 {
		return EvaluateOutput{Decision: PayloadHold, Reason: "judge returned invalid confidence", Usage: usage}, fmt.Errorf("confidence %f out of [0,1] range", args.Confidence)
	}

	out := EvaluateOutput{Reason: args.Reason, Usage: usage}
	switch {
	case args.Confidence < c.confidenceThreshold:
		out.Decision = PayloadHold
	case args.Safe:
		out.Decision = PayloadAllow
	default:
		out.Decision = PayloadBlock
	}
	return out, nil
}

func formatJudgeUserMessage(in EvaluateInput) string {
	objective := in.Objective
	if objective == "" {
		objective = "(none)"
	}
	reason := in.AgentReason
	if reason == "" {
		reason = "(none)"
	}
	body := string(in.Body)
	if len(body) > 4096 {
		body = body[:4096] + "\n[truncated]"
	}
	return fmt.Sprintf(
		"vuln_class: %s\ncage_type: %s\nmethod: %s\nurl: %s\nobjective: %s\nagent_reason: %s\nbody:\n%s",
		in.VulnClass, in.CageType, in.Method, in.URL, objective, reason, body,
	)
}

File: sdk/typescript/examples/webhook/judge.ts
/**
 * Judge handler — passthrough to an OpenAI-compatible chat completion
 * provider.
 *
 * Agentcage's JudgeClient builds the request (system + user prompts,
 * tools[]: [submit_judgment], tool_choice: "required") and parses the
 * response. This handler is a thin forwarder: receives the chat
 * completion, overrides model with JUDGE_MODEL, sends to the provider,
 * returns the response verbatim.
 *
 * Want custom behavior (extra context, provider switching, per-customer
 * tuning)? Fork this file and modify the request before forwarding.
 * The wire contract is OpenAI chat completions in, OpenAI chat
 * completions out — keep the `tools[]` + `tool_choice` fields intact
 * and agentcage's parsing will continue to work. Remove them and
 * agentcage will treat every request as malformed → human review.
 *
 * Environment:
 *   JUDGE_PROVIDER_URL  LLM endpoint (default: same as LLM_PROVIDER_URL)
 *   JUDGE_PROVIDER_KEY  API key (default: LLM_PROVIDER_KEY)
 *   JUDGE_MODEL         Model name (default: gpt-4o-mini — cheaper than
 *                       LLM_MODEL since judge runs on every flagged request)
 *
 * agentcage config:
 *   judge:
 *     endpoint: "https://your-webhook-host:8082/judge"
 *
 *   Store the API key in Vault (same value as the LLM key, since both
 *   routes share WEBHOOK_API_KEY):
 *     agentcage vault put orchestrator/judge-api-key value=<WEBHOOK_API_KEY>
 */

import { IncomingMessage, ServerResponse } from 'node:http';
import { readBody, sendJSON } from './index';

const PROVIDER_URL = process.env.JUDGE_PROVIDER_URL
  ?? process.env.LLM_PROVIDER_URL
  ?? 'https://api.openai.com/v1/chat/completions';
const PROVIDER_KEY = process.env.JUDGE_PROVIDER_KEY || process.env.LLM_PROVIDER_KEY || '';
const MODEL = process.env.JUDGE_MODEL ?? 'gpt-4o-mini';

if (!PROVIDER_KEY) {
  console.error('JUDGE_PROVIDER_KEY (or LLM_PROVIDER_KEY as fallback) is required');
  process.exit(1);
}

export async function handleJudge(req: IncomingMessage, res: ServerResponse) {
  if (req.method !== 'POST') {
    sendJSON(res, 405, { error: 'POST required' });
    return;
  }

  let body: any;
  try {
    const raw = await readBody(req);
    body = JSON.parse(raw.toString());
  } catch {
    sendJSON(res, 400, { error: 'invalid JSON' });
    return;
  }

  // Webhook decides the model so operators can use a cheaper model for
  // judge than for the main LLM. Caller can request a model but we
  // override — same pattern as the /llm route.
  body.model = MODEL;

  try {
    const providerResp = await fetch(PROVIDER_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${PROVIDER_KEY}`,
      },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(30_000),
    });

    const providerBody = await providerResp.text();
    res.writeHead(providerResp.status, { 'Content-Type': 'application/json' });
    res.end(providerBody);
  } catch (err: any) {
    console.error('judge provider request failed:', err.message);
    sendJSON(res, 502, { error: `provider unreachable: ${err.message}` });
  }
}
